### ISSUE TITLE
Script editor scroll too fast with trackpad in OSX.

### ISSUE THREAD

**Issue by lizistorm** [godotengine/godot#3913]
**Operating system or device:**
OSX, trackpad

**Issue description:**
GDScript editor scroll too fast vertical and too slow horizontal with trackpad. It hard to scroll & view with long script file or help file.

**Comment by GalanCM** [comment]
I can verify this, and it's quite frustrating. For added confusion, horizontal scrolling is inverted, but vertical is not.

**Comment by Hinsbart** [comment]
Horizontal scrolling is inverted, really? I added this feature but couldn't test it as I don't have a mac.
However in #3565, someone told me it's working, mhh...
Could it be that the trackpad behaves differently than the mac mouse?

**Comment by kirilledelman** [comment]
I said it was working. But he's right - horizontal scrolling is backwards.
I also agree that trackpad scrolling on a mac is insanely fast / too sensitive. You move your finger a millimeter and it scrolls 10 lines. Is there a way to make scroll distance factor into number of lines scrolled?

**Comment by Hinsbart** [comment]
Allright. So for the record, vertical scrolling is too fast and horizontal scrolling is too slow and inverted?

**Comment by Hinsbart** [comment]
Ok, the inverted scrolling should be fixed by #3972 .

We'll have to see how to fix the scroll speed... the platform code just sends `MOUSE_WHEEL_*` events and the Script editor then recieves them and handles the scrolling. IIRC, the speed is 3 (chars/lines?) per event.
But it wouldn't be a good idea to introduce platform-specific code in the text-editor. So what about making the scroll speed customizable in the editor settings? The OSX platform could then set appropriate defaults for it.

**Comment by lizistorm** [comment]
That way will affect mouse wheel and trackpad at the same time. Current mouse wheel speed is OK. But if trackpad slow to normal speed the speed of mouse wheel will be too slow then. Are there an event like TRACKPAD_MOTION separately?

**Comment by Hinsbart** [comment]
Mhh.. I don't think so, but currently the event is sent every time the `(void)scrollWheel:(NSEvent *)` callback is called. There's a `delta` value which I guess holds information on how much has been scrolled. Maybe we could introduce some kind of deadzone here.

But I don't have a mac, so someone else would have to look into this.
I'd start by printing out `deltaX` and `deltaY`, in `platform/osx/os_osx.mm` (line 796+) and comparing the trackpad output to the mouse wheel.

**Comment by kirilledelman** [comment]
As a Mac user, I'd be happy with an editor setting for scroll speed. No platform-dependent code necessary. It would probably be a good idea to have horizontal and vertical scrolling speed be separate. Unfortunately I'm not experienced in compiling the engine, but maybe I should try setting it up to help out.

**Comment by toger5** [comment]
I have set everything up for testing so ill do some test and tell you what I got.
I think the best approach would be that we add another param to the mouse scroll godot event: `value`
than we can use default values for each operating system and apply them to the different ui elements.
of course that would be a lot of work to implement for each and every element that uses scrolling.
But it would allow for really nice trackpad control.
I also would propose to add: a zoom_in and zoom_out event (but I guess I should open another issue for that...)
